# market_data.proto
syntax = "proto3";

package marketdata;

service MarketDataService {
    // Bidirectional streaming RPC
    rpc StreamMarketData (stream SubscriptionRequest) returns (stream MarketDataResponse) {}
}

message SubscriptionRequest {
    int32 instrument_id = 1;
    enum RequestType {
        SUBSCRIBE = 0;
        UNSUBSCRIBE = 1;
    }
    RequestType type = 2;
}

message MarketDataResponse {
    int32 instrument_id = 1;
    enum UpdateType {
        SNAPSHOT = 0;
        INCREMENT = 1;
        CLEAR = 2;  // For unsubscribe notifications
    }
    UpdateType type = 2;
    OrderBookData data = 3;
}

message OrderBookData {
    repeated PriceLevel bids = 1;
    repeated PriceLevel asks = 2;
}

message PriceLevel {
    double price = 1;
    int32 quantity = 2;
}